### PR TITLE
Replace memory batch size with a fraction of the total batch size

### PR DESCRIPTION
**Pull request by lballes**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/renate/updaters/experimental/er.py [hunk 1]
```diff
@@ -95,7 +95,7 @@ def train_dataloader(self) -> DataLoader:
         )
         return DataLoader(
             train_dataset,
-            batch_size=self._batch_size,
```
Comment (alexanderbuchholz) [inline src/renate/updaters/experimental/er.py:98]: Nit: Would it make sense to add a check for the value of `self._batch_memory_frac`? 
I.e., add as statement of the form 
`assert 0 <= self._batch_memory_frac <= 1.`

File: src/renate/updaters/experimental/er.py [hunk 2]
```diff
@@ -192,7 +192,7 @@ def training_step(
 
     def _sample_from_buffer(self, device: torch.device) -> Optional[Tuple[NestedTensors, DataDict]]:
         """Function to sample from the buffer, if buffer is populated."""
-        if self._memory_loader is not None and len(self._memory_buffer) >= self._memory_batch_size:
```
Comment (alexanderbuchholz) [inline src/renate/updaters/experimental/er.py:195]: Is changing this statement safe? This could create a different behavior,  no?

File: src/renate/updaters/experimental/gdumb.py [hunk 3]
```diff
@@ -44,6 +44,7 @@ def __init__(
     ) -> None:
         super().__init__(
             memory_size=memory_size,
+            batch_memory_frac=-1.0,  # Dummy value
```
Comment (wistuba) [inline src/renate/updaters/experimental/gdumb.py:47]: why not 1?

Comment (alexanderbuchholz) [inline src/renate/updaters/experimental/gdumb.py:47]: Should this raise an error?

File: src/renate/updaters/experimental/repeated_distill.py [hunk 4]
```diff
@@ -229,7 +229,7 @@ def __init__(
         super().__init__(
             model=model,
             memory_size=memory_size,
-            memory_batch_size=batch_size,
+            batch_memory_frac=-1.0,
```
Comment (wistuba) [inline src/renate/updaters/experimental/repeated_distill.py:232]: is it not using the memory?

File: src/renate/updaters/experimental/offline_er.py [hunk 5]
```diff
@@ -66,17 +65,32 @@ def on_model_update_start(
         self._num_points_current_task = len(train_dataset)
 
     def train_dataloader(self) -> DataLoader:
-        train_loader = super().train_dataloader()
-        loaders = {"current_task": train_loader}
-        if len(self._memory_buffer) > self._memory_batch_size:
+        """Returns the dataloader for training the model."""
+        loaders = {}
+        memory_batch_size = int(self._batch_memory_frac * self._batch_size)
+        if len(self._memory_buffer) >= memory_batch_size:
```
Comment (alexanderbuchholz) [inline src/renate/updaters/experimental/offline_er.py:71]: Not directly related to this PR, but would it make sense to add a check somewhere that the buffer size must always be larger than the memory size? Otherwise the memory is always ignored.

File: examples/benchmarking/class_incremental_learning_cifar10_der.py [hunk 6]
```diff
@@ -11,8 +11,8 @@
     "learning_rate": 0.03,
     "alpha": 0.2,
     "beta": 0.5,
-    "batch_size": 32,
-    "memory_batch_size": 32,
+    "batch_size": 64,
+    "batch_memory_frac": 0.5,
```
Comment (wistuba) [inline examples/benchmarking/class_incremental_learning_cifar10_der.py:15]: I'd prefer full words over abbreviations

File: src/renate/updaters/learner.py [hunk 7]
```diff
@@ -404,14 +402,14 @@ class ReplayLearner(Learner, abc.ABC):
...
-        memory_batch_size: int = defaults.BATCH_SIZE,
+        batch_memory_frac: float = defaults.BATCH_MEMORY_FRAC,
         buffer_transform: Optional[Callable] = None,
         buffer_target_transform: Optional[Callable] = None,
         seed: int = defaults.SEED,
         **kwargs,
     ) -> None:
         super().__init__(seed=seed, **kwargs)
-        self._memory_batch_size = min(memory_size, memory_batch_size)
+        self._batch_memory_frac = min(batch_memory_frac, memory_size / self._batch_size)
```
Comment (alexanderbuchholz) [inline src/renate/updaters/learner.py:412]: and here we could check that `memory_size > batch_memory_frac * self._batch_size`

File: test/conftest.py [hunk 8]
```diff
@@ -113,7 +113,7 @@ def pytest_collection_modifyitems(config, items):
     },
     OfflineExperienceReplayLearner: {
         "memory_size": 30,
-        "memory_batch_size": 20,
+        "batch_memory_frac": 0.6,
```
Comment (alexanderbuchholz) [inline test/conftest.py:116]: What is the `batch_size` here? if the total batch size is 30 and the memory batch size is 20 this implies `batch_memory_frac = 0.666666`. But the values here would imply `batch_size = 33.33333`. (Or is it me doing the computation wrong here?). Thanks

File: test/integration_tests/configs/suites/quick/offline-er.json [hunk 9]
```diff
@@ -5,6 +5,6 @@
     "dataset": "cifar10.json",
     "backend": "local",
     "job_name": "class-incremental-mlp-offline-er",
-    "expected_accuracy": [0.7089999914169312, 0.5120000243186951],
```
Comment (alexanderbuchholz) [inline test/integration_tests/configs/suites/quick/offline-er.json:8]: why is the expected accuracy changing? This should not be the case, no?

**Review by alexanderbuchholz — COMMENTED** [review]
Thanks for looking into this! Please see my comments below.

File: test/conftest.py (hunk 8 above)
Comment (wistuba) [reply test/conftest.py:116]: true, this should imply that memory_batch_size will be 18

File: test/integration_tests/configs/suites/quick/offline-er.json (hunk 9 above)
Comment (lballes) [reply test/integration_tests/configs/suites/quick/offline-er.json]: The behavior of Offline-ER is changed slightly by this PR. In the first update, it will now use the total batch size exclusively for the current task data, since no memory is available yet. Hence, small changes are to be expected and I updated the numbers here.

File: test/conftest.py (hunk 8 above)
Comment (lballes) [reply test/conftest.py:116]: I didn't pay close attention to the numbers, since the tests don't depend on exact memory sizes.

File: src/renate/updaters/experimental/er.py (hunk 1 above)
Comment (lballes) [reply src/renate/updaters/experimental/er.py:98]: Yes, good point. Will add this.

File: src/renate/updaters/experimental/er.py [hunk 10]
```diff
@@ -192,7 +192,7 @@ def training_step(
...
+        if self._memory_loader is not None:
```
[the remark this answers is not in the document]
Comment (lballes) [reply src/renate/updaters/experimental/er.py:195]: Will add a check.

File: src/renate/updaters/experimental/er.py [hunk 11]
```diff
@@ -219,10 +219,11 @@ def _update_memory_buffer(self, step_output: STEP_OUTPUT) -> None:
 
     def _set_memory_loader(self) -> None:
         """Create a memory loader from a memory buffer."""
-        if self._memory_loader is None and len(self._memory_buffer) >= self._memory_batch_size:
+        memory_batch_size = int(self._batch_memory_frac * self._batch_size)
+        if self._memory_loader is None and len(self._memory_buffer) >= memory_batch_size:
```
[the remark this answers is not in the document]
Comment (lballes) [reply src/renate/updaters/experimental/er.py:223]: Because this is creating the memory loader, so we want to make sure it will return at least one batch. In `_sample_memory_buffer` it should then suffice to check for the existence of a memory loader, so I removed the superfluous second part.

File: src/renate/updaters/experimental/gdumb.py (hunk 3 above)
Comment (lballes) [reply src/renate/updaters/experimental/gdumb.py:47]: Will change that.

File: src/renate/updaters/experimental/repeated_distill.py (hunk 4 above)
Comment (lballes) [reply src/renate/updaters/experimental/repeated_distill.py:232]: It is *only* sampling from the memory using the entire batch size. So passing a dummy value to the parent class (which probably means this inheritance structure is not ideal...)

File: src/renate/updaters/experimental/gdumb.py (hunk 3 above)
Comment (lballes) [reply src/renate/updaters/experimental/gdumb.py:47]: Will add a check for to be in [0, 1] and then use 0.0 as a dummy value.

File: src/renate/updaters/learner.py (hunk 7 above)
Comment (lballes) [reply src/renate/updaters/learner.py:412]: Yes, I'll add that.

File: src/renate/updaters/experimental/gdumb.py [hunk 12]
```diff
@@ -45,9 +45,9 @@ def __init__(
         super().__init__(
             memory_size=memory_size,
             seed=seed,
+            # batch_memory_frac=0.0,
```
Comment (prabhuteja12) [inline src/renate/updaters/experimental/gdumb.py:48]: is this unused?

File: test/conftest.py [hunk 13]
```diff
@@ -69,7 +69,7 @@ def pytest_collection_modifyitems(config, items):
 LEARNER_KWARGS = {
     ExperienceReplayLearner: {
         "memory_size": 30,
-        "memory_batch_size": 20,
+        "batch_memory_frac": 0.6,
```
Comment (prabhuteja12) [inline test/conftest.py:72]: The ratios dont match: .6 * 30 != 20?

Comment (lballes) [reply test/conftest.py:72]: Yes, will adjust this. But the tests don't depend on the exact numbers. All unit tests are passing. It's the integration tests that puzzle me.

File: src/renate/updaters/experimental/offline_er.py [hunk 14]
```diff
@@ -77,12 +76,28 @@ def on_model_update_start(
...
+            transform=self._train_transform,
+            target_transform=self._train_target_transform,
+        )
+        loaders = {
+            "current_task": DataLoader(
+                train_dataset,
+                batch_size=self._batch_size - memory_batch_size,
+                shuffle=True,
+                generator=self._rng,
+                pin_memory=True,
```
Comment (wistuba) [inline src/renate/updaters/experimental/offline_er.py:94]: was this the PR where you wanted to add the drop_last argument?

File: src/renate/updaters/avalanche/learner.py [hunk 15]
```diff
@@ -73,10 +73,11 @@ class AvalancheReplayLearner(ReplayLearner, AvalancheLoaderMixin):
     def create_avalanche_learner(
         self, plugins: List[SupervisedPlugin], **kwargs: Any
     ) -> BaseSGDTemplate:
+        memory_batch_size = round(self._batch_memory_frac * self._batch_size)
```
Comment (wistuba) [inline src/renate/updaters/avalanche/learner.py:76]: you could simply reuse `self._memory_batch_size`

File: src/renate/updaters/model_updater.py [hunk 16]
```diff
@@ -411,8 +411,8 @@ def _fit_learner(
 
         strategy = create_strategy(self._devices, self._strategy)
         # Finetuning-equivalent epochs.
-        num_batches = len(learner._train_dataset) // learner._batch_size
-        num_batches += min(len(learner._train_dataset) % learner._batch_size, 1)
+        num_batches = len(learner._train_dataset) // 256  # learner._batch_size
+        num_batches += min(len(learner._train_dataset) % 256, 1)  # learner._batch_size, 1)
```
Comment (wistuba) [inline src/renate/updaters/model_updater.py:415]: not sure why this doesn't work on github, but it works locally. the reason the tests are failing is that a pass over the data has now effectively fewer batches.

what's a bit concerning is that I saw that the validation loss has changed. are there any changes to how the validation buffer gets populated or do we drop the last batch of the validation buffer? funnily, that does not happen for each epoch.
